cpp/cmake/thirdparty/get_libcudacxx.cmake

```diff
@@ -0,0 +1,37 @@
+# =============================================================================
+# Copyright (c) 2023, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+# in compliance with the License. You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under the License
+# is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+# or implied. See the License for the specific language governing permissions and limitations under
+# the License.
+# =============================================================================
+
+# This function finds thrust and sets any additional necessary environment variables.
+function(find_and_configure_libcudacxx)
+  # Make sure we install libcudacxx beside our patched version of thrust
+  include(GNUInstallDirs)
+  set(CMAKE_INSTALL_INCLUDEDIR "${CMAKE_INSTALL_INCLUDEDIR}/libcuspatial")
+  set(CMAKE_INSTALL_LIBDIR "${CMAKE_INSTALL_INCLUDEDIR}/lib")
```

Comment:
Hmm I realize that this is copied directly from libcudf, but something seems off here. Why are we setting the libdir to be inside the include dir? If that's to have the side-by-side behavior, then I suspect what we really want is perhaps `"${CMAKE_INSTALL_INCLUDEDIR}/libcuspatial" too. @robertmaynard could you verify this?

I suspect we're getting away with an error here because libcudacxx is header-only so there aren't any libraries to install in practice. The only other possibility is that this lib dir setting is correct and necessary in order for the CMake config files to be installed in the normal search directories (lib/cmake), but that seems like a fragile approach if libcudacxx does ever start shipping libraries.

If my diagnosis is right we may need to port this fix back to libcudf too.

Also, if libcuspatial is doing the exact same patching as libcudf perhaps we could reuse libcudf's copy instead of generating yet another one? But I can see the possibility for conflict there so I'm OK with this approach too.

Reply:
> Why are we setting the libdir to be inside the include dir? I

Cudf does so since it is a patched version of the CCCL libraries and therefore shouldn't be picked up by any abitrary `find_package(Thrust)` since cudf disables some thrust features that other libraries could need.

Second, since we are constrained on the `CMAKE_INSTALL_INCLUDEDIR` and `CMAKE_INSTALL_LIBDIR` locations since the CCCL find packages search for headers and pre 2.2 had issues when the `include` and `lib` didn't share the same root ( `CMAKE_INSTALL_INCLUDEDIR` in this case ).

Reply:
@isVoid We should double-check if we get libcudf's patched Thrust if we remove the changes for fetching/patching Thrust here, now that the cuda-nvcc include flag order has been fixed.

Reply:
I get that we're trying to avoid normal `find_package` calls from finding the library, but I guess I would have expected us to want to place the libraries in some subdirectory of the lib dir (with a different name to subvert normal `find_package(Thrust)`) rather than putting libraries into the _include_ dir. Using a top-level `$INCLUDEDIR/lib` also seems fairly clobber-friendly, wouldn't it be better to place it inside a `rapids` or `cuspatial` directory?

Reply:
> now that the cuda-nvcc include flag order has been fixed.

You mean using -I in nvcc compiler instead of -isystem? Can you post the link to this fix for track record?

Reply:
It moving the implicit include of -I$PREFIX/targets/… to the end rather than the beginning of the search path. No changes in -I vs. -isystem. https://github.com/conda-forge/cuda-nvcc-feedstock/pull/22

Reply:
> We should double-check if we get libcudf's patched Thrust if we remove the changes for fetching/patching Thrust here, now that the cuda-nvcc include flag order has been fixed.

It seems removing the patched thrust now no longer causes the issue.

Reply:
As long as this is working, given that it's copied from libcudf I'm fine approving here and following up with Robert separately. Addressing my concerns here is not on the critical path for CUDA 12.